### PR TITLE
Code base wide cleanup (mostly C++17 related)

### PR DESCRIPTION
**Pull request by anstaf**
- `GT_CONSTEXPR` and `GT_CONSTEXPR_TARGET` goes away
- `wstd` stuff goes away
- `is_trivially_copy_constructible` check is consistently used instead of `is_trivially_copyable` where the data is passed host/device boundary, because it is exactly what is needed.
- `make_smth` pattern is replaced to template argument deduction in several places
- `composite` is rewritten using c++17
- `overload` is rewritten using c++17
-  `std::smth_v<...>` are used instead of `std::smth<...>::value`
- `static_assert(<cond>)` used instead of `static_assert(<cond>, "")`

### CODE REVIEW

**Comment by anstaf** [comment]
launch jenkins

**Comment by anstaf** [comment]
launch jenkins

**Comment by anstaf** [comment]
launch jenkins

**Comment by anstaf** [comment]
launch perftest

File: include/gridtools/common/cuda_util.hpp [hunk 1]
```diff
@@ -53,21 +53,21 @@ namespace gridtools {
...
-        template <class T, std::enable_if_t<!std::is_array<T>::value, int> = 0>
+        template <class T, std::enable_if_t<!std::is_array_v<T>, int> = 0>
         unique_cuda_ptr<T> cuda_malloc() {
             T *ptr;
             GT_CUDA_CHECK(cudaMalloc(&ptr, sizeof(T)));
             return unique_cuda_ptr<T>{ptr};
         }
 
-        template <class T, std::enable_if_t<std::is_trivially_copyable<T>::value, int> = 0>
+        template <class T, std::enable_if_t<std::is_trivially_copy_constructible_v<T>, int> = 0>
```
Comment (fthaler) [inline include/gridtools/common/cuda_util.hpp:63]: Is the change from `is_trivially_copyable` to `is_trivially_copy_constructible` intended?

File: include/gridtools/common/cuda_util.hpp [hunk 2]
```diff
@@ -53,21 +53,21 @@ namespace gridtools {
...
         unique_cuda_ptr<T> make_clone(T const &src) {
             unique_cuda_ptr<T> res = cuda_malloc<T>();
             GT_CUDA_CHECK(cudaMemcpy(res.get(), &src, sizeof(T), cudaMemcpyHostToDevice));
             return res;
         }
 
-        template <class T, std::enable_if_t<std::is_trivially_copyable<T>::value, int> = 0>
+        template <class T, std::enable_if_t<std::is_trivially_copy_constructible_v<T>, int> = 0>
```
Comment (fthaler) [inline include/gridtools/common/cuda_util.hpp:70]: Same here.

File: include/gridtools/stencil/global_parameter.hpp [hunk 3]
```diff
@@ -22,23 +22,26 @@ namespace gridtools {
 
             template <class T>
             struct global_parameter {
-                static_assert(std::is_trivially_copyable<T>(), "global parameter should be trivially copyable");
+                static_assert(std::is_trivially_copy_constructible_v<T>, "global parameter should be trivially copyable");
```
Comment (fthaler) [inline include/gridtools/stencil/global_parameter.hpp:25]: Again `trivially_copyable` vs. `trivially_copy_constructible`. If this intended, the comment should probably be adapted, too.

File: include/gridtools/stencil/gpu/launch_kernel.hpp [hunk 4]
```diff
@@ -142,7 +142,7 @@ namespace gridtools {
                     static_assert(Extent::iminus::value <= 0, GT_INTERNAL_ERROR);
                     static_assert(Extent::iplus::value >= 0, GT_INTERNAL_ERROR);
 
-                    static_assert(std::is_trivially_copyable<Fun>::value, GT_INTERNAL_ERROR);
+                    static_assert(std::is_trivially_copy_constructible_v<Fun>, GT_INTERNAL_ERROR);
```
Comment (fthaler) [inline include/gridtools/stencil/gpu/launch_kernel.hpp:145]: Again `trivially_copyable` vs. `trivially_copy_constructible`.

File: include/gridtools/stencil/gpu/launch_kernel.hpp [hunk 5]
```diff
@@ -171,7 +171,7 @@ namespace gridtools {
                     std::enable_if_t<is_empty_ij_extents<Extent>(), int> = 0>
                 void launch_kernel(int_t i_size, int_t j_size, uint_t zblocks, Fun fun, size_t shared_memory_size = 0) {
 
-                    static_assert(std::is_trivially_copyable<Fun>::value, GT_INTERNAL_ERROR);
+                    static_assert(std::is_trivially_copy_constructible_v<Fun>, GT_INTERNAL_ERROR);
```
Comment (fthaler) [inline include/gridtools/stencil/gpu/launch_kernel.hpp:174]: Again `trivially_copyable` vs. `trivially_copy_constructible`.

File: include/gridtools/stencil/gpu_horizontal/launch_kernel.hpp [hunk 6]
```diff
@@ -55,7 +55,7 @@ namespace gridtools {
 
                 template <int_t BlockSizeI, int_t BlockSizeJ, class Fun>
                 void launch_kernel(int_t i_size, int_t j_size, uint_t zblocks, Fun fun) {
-                    static_assert(std::is_trivially_copyable<Fun>::value, GT_INTERNAL_ERROR);
+                    static_assert(std::is_trivially_copy_constructible_v<Fun>, GT_INTERNAL_ERROR);
```
Comment (fthaler) [inline include/gridtools/stencil/gpu_horizontal/launch_kernel.hpp:58]: Again `trivially_copyable` vs. `trivially_copy_constructible`.

File: include/gridtools/storage/adapter/python_sid_adapter.hpp [hunk 7]
```diff
@@ -113,8 +113,8 @@ namespace gridtools {
 
         template <class T, std::size_t Dim, class Kind = void, size_t UnitStrideDim = size_t(-1)>
         wrapper<T, Dim, Kind, UnitStrideDim> as_sid(pybind11::buffer const &src) {
-            static_assert(
-                std::is_trivially_copyable<T>::value, "as_sid should be instantiated with the trivially copyable type");
+            static_assert(std::is_trivially_copy_constructible_v<T>,
```
Comment (fthaler) [inline include/gridtools/storage/adapter/python_sid_adapter.hpp:116]: Again `trivially_copyable` vs. `trivially_copy_constructible`.

File: include/gridtools/storage/adapter/python_sid_adapter.hpp [hunk 8]
```diff
@@ -266,7 +266,7 @@ namespace gridtools {
 
         template <class T, size_t Dim, class Kind = void, size_t UnitStrideDim = size_t(-1)>
         auto as_cuda_sid(pybind11::object const &src) {
-            static_assert(std::is_trivially_copyable<T>::value,
+            static_assert(std::is_trivially_copy_constructible_v<T>,
```
Comment (fthaler) [inline include/gridtools/storage/adapter/python_sid_adapter.hpp:269]: Again `trivially_copyable` vs. `trivially_copy_constructible`.

File: tests/include/cuda_test_helper.hpp [hunk 9]
```diff
@@ -30,7 +30,7 @@ namespace gridtools {
         }
         template <class T, T (*Fun)()>
         T exec_with_shared_memory(size_t shm_size, integral_constant<T (*)(), Fun> fun) {
-            static_assert(std::is_trivially_copyable<T>::value, "");
+            static_assert(std::is_trivially_copy_constructible_v<T>);
```
Comment (fthaler) [inline tests/include/cuda_test_helper.hpp:33]: Again `trivially_copyable` vs. `trivially_copy_constructible`.

File: tests/include/cuda_test_helper.hpp [hunk 10]
```diff
@@ -39,22 +39,23 @@ namespace gridtools {
...
         }
         template <class Fun, class... Args>
         auto exec_with_shared_memory(size_t shm_size, Fun fun, Args... args) {
-            static_assert(!std::is_pointer<Fun>::value, "");
-            static_assert(std::conjunction<std::negation<std::is_pointer<Args>>...>::value, "");
+            static_assert(!std::is_pointer_v<Fun>);
+            static_assert(std::conjunction_v<std::negation<std::is_pointer<Args>>...>);
             using res_t = std::decay_t<decltype(fun(args...))>;
-            static_assert(std::is_trivially_copyable<res_t>::value, "");
+            static_assert(std::is_trivially_copy_constructible_v<res_t>);
```
Comment (fthaler) [inline tests/include/cuda_test_helper.hpp:50]: Again `trivially_copyable` vs. `trivially_copy_constructible`.

File: tests/unit_tests/stencil/core/test_functor_metafunctions.cpp [hunk 11]
```diff
@@ -19,13 +19,13 @@ namespace gridtools {
                 using interval_t = interval<level<0, 1, 3>, level<2, -1, 3>>;
 
                 struct empty {};
-                //                static_assert(check_valid_apply_overloads<empty, interval_t>::value, "");
+                //                static_assert(check_valid_apply_overloads<empty, interval_t>::value);
```
Comment (fthaler) [inline tests/unit_tests/stencil/core/test_functor_metafunctions.cpp:22]: Why those commented assertions (also below)? Are those cases that _should_ work but actually don’t with the currently implemented algorithm?

**Review by fthaler — CHANGES_REQUESTED** [review]

File: tests/unit_tests/stencil/core/test_functor_metafunctions.cpp (hunk 11 above)
Comment (anstaf) [reply tests/unit_tests/stencil/core/test_functor_metafunctions.cpp:22]: Explanation comment is added

File: include/gridtools/boundaries/distributed_boundaries.hpp [hunk 12]
```diff
@@ -139,7 +139,7 @@ namespace gridtools {
                 \param jobs Variadic list of jobs
             */
             template <typename... Jobs>
-            void boundary_only(Jobs const &... jobs) {
+            void boundary_only(Jobs const &...jobs) {
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/boundaries/distributed_boundaries.hpp:142]: Most likely it is clang-format changes.  I am not motivated to waist time on it now. Let us let as it is in this PR.

File: include/gridtools/common/cuda_util.hpp (hunk 1 above)
Comment (anstaf) [reply include/gridtools/common/cuda_util.hpp:63]: yes. explained in PR description

File: include/gridtools/common/cuda_util.hpp (hunk 2 above)
Comment (anstaf) [reply include/gridtools/common/cuda_util.hpp:70]: same

File: include/gridtools/common/cuda_util.hpp (hunk 1 above)
Comment (fthaler) [reply include/gridtools/common/cuda_util.hpp:63]: Ups, too boring review, I even missed the description. Sorry!

File: tests/unit_tests/stencil/core/test_functor_metafunctions.cpp (hunk 11 above)
Comment (fthaler) [reply tests/unit_tests/stencil/core/test_functor_metafunctions.cpp:30]: Thanks, that’s helpful indeed.

**Review by fthaler — APPROVED** [review]

**Comment by anstaf** [comment]
launch perftest

**Comment by anstaf** [comment]
launch jenkins

**Pull request merged**